libavcodec/libproresdecoder.c

```diff
@@ -0,0 +1,237 @@
+                    memcpy(framePtr->data[1] + i * framePtr->linesize[1] + (j / (4*dataUnitBytes))*dataUnitBytes, dataPtr + i * rowBytes + j, dataUnitBytes);
+                }
+                else if (j % (2*dataUnitBytes) == 0) { // V component.
+                    memcpy(framePtr->data[2] + i * framePtr->linesize[2] + (j / (4*dataUnitBytes))*dataUnitBytes, dataPtr + i * rowBytes + j, dataUnitBytes);
+                }
+                else if (j % dataUnitBytes == 0) { // Y component
+                    memcpy(framePtr->data[0] + i * framePtr->linesize[0] + (j / (2*dataUnitBytes))*dataUnitBytes, dataPtr + i * rowBytes + j, dataUnitBytes);
+                }
+            }
+        }
+    } else if (decoderctx->pix_fmt == kPRFormat_y416) {
+        memset(framePtr->data[3], 0, framePtr->linesize[0] * decoderctx->m_height); // A
+        // convert the y416 layout [A0 - 2B LE, Y0-2B LE, Cb0-2B LE, Cr0-2B LE] to YUVA444p16LE (planar YCbCrA) 
+        int dataUnitBytes = 2; // read 2 bytes at a time
+        for (int i = 0; i < decoderctx->m_height; i++) {
+            for (int j = 0; j < rowBytes; j=j+dataUnitBytes) {
+                int op_plane = ((j+6)%8)/2; // We want to right shift the AYCbCr order by 3 and round to get YCbCrA order
+                int write_byte = (j / (4*dataUnitBytes))*dataUnitBytes; // 4 pixels = 4*dataUnitBytes bytes and write dataUnitBytes at a time
+                memcpy(framePtr->data[op_plane] + i * framePtr->linesize[op_plane] + write_byte, dataPtr + i * rowBytes + j, dataUnitBytes);
+            }
+        }
+    }
+    free(dataPtr);
+    return res;
+};
+
+#if HAVE_THREADS
+static int update_thread_context(AVCodecContext *dstavctx, const AVCodecContext *srcavctx)
+{
+    ProResDecodeCodec *srcdecodectx = srcavctx->priv_data;
+    ProResDecodeCodec *dstdecodectx = dstavctx->priv_data;
+     
+    dstavctx->profile = srcavctx->profile;
+    dstavctx->bits_per_raw_sample = srcavctx->bits_per_raw_sample;
+    dstavctx->pix_fmt = srcavctx->pix_fmt;
+
+    dstdecodectx->m_width = srcdecodectx->m_width; 
+    dstdecodectx->m_height = srcdecodectx->m_height;
+    dstdecodectx->pix_fmt = srcdecodectx->pix_fmt;
+
+    // For now the downscale mode is universally 1x1
+    dstdecodectx->downscale_mode = kPRDownscale_1x1;
+
+    return 0;
+};
+#endif
+
+int closeAppleProResDecoder(struct AVCodecContext* avctx) {
+    ProResDecodeCodec* proresdecodecctx = avctx->priv_data;
+    PRCloseDecoder(proresdecodecctx->m_decoderContext);
+    return 0;
+};
+
+const FFCodec ff_libproresdecoder_decoder = {
+    .p.name         = "libproresdecoder",
+    .p.long_name    = NULL_IF_CONFIG_SMALL("Apple ProRes (libProResDecoder.a)"),
+    .p.type         = AVMEDIA_TYPE_VIDEO,
+    .p.id           = AV_CODEC_ID_PRORES,
+    .priv_data_size = sizeof(ProResDecodeCodec),
+    .init           = initAppleProResDecoder,
+    .close          = closeAppleProResDecoder,
+    FF_CODEC_DECODE_CB(decodeAppleProResFrame),
```

Comment:
The 3 main functions to this interface are the ones that `init`, `close`, `decode`